docs/TUTORIAL.md

```diff
@@ -1,3 +1,413 @@
+
+// call the modified object that now includes the deployment and the service
+k.core.v1.list.new([ingressController, ingressService])
+```
+
+Save the file again, and run again:
+
+```bash
+jsonnet ingressController.jsonnet > deployment.yaml
+kubectl apply -f deployment.yaml
+```
+
+The generated YAML now looks like this:
+
+```yaml
+apiVersion: v1
+items:
+  - apiVersion: extensions/v1beta1
+    kind: Deployment
+    metadata:
+      labels:
+        run: nginx-ingress-controller
+      name: nginx-ingress-controller
+      namespace: nginx-ingress
+    spec:
+      replicas: 2
+      template:
+        metadata:
+          labels:
+            run: nginx-ingress-controller
+        spec:
+          containers:
+            - image: 'gcr.io/google_containers/nginx-ingress-controller:0.9.0-beta.7'
+              name: nginx-ingress-controller
+              ports:
+                - containerPort: 80
+                  hostPort: 80
+                - containerPort: 443
+                  hostPort: 443
+          terminationGracePeriodSeconds: 60
+  - apiVersion: v1
+    kind: Service
+    metadata:
+      labels:
+        run: nginx-ingress-controller
+      name: nginx-ingress-controller
+      namespace: nginx-ingress
+    spec:
+      ports:
+        - name: http
+          port: 80
+        - name: https
+          port: 443
+      selector:
+        run: nginx-ingress-controller
+      type: LoadBalancer
+kind: List
+```
+
+You can also easily add the built-in mixins for liveness and readiness probes to the container object.
+
+## Write a custom mixin to include as a sidecar
+
+Your application needs logging and monitoring, so you can write a mixin to define a 
+Fluentd container. You can then add it to your cluster as a sidecar.
+
+Because mixins let you assemble components at runtime (dynamic inheritance),
+this approach means that different team members can define the 
+Kubernetes objects that your cluster needs. You can then mix them into 
+your master configuration without having to copy all the details.
+
+Fluentd runs in a DaemonSet, so let's start with this definition: 
+
+```javascript
+    daemonSet(name, containerName, containerTag, namespace)::
+      ds.new() +
+      // Metadata.
+      ds.mixin.metadata.name(name) +
+      ds.mixin.metadata.namespace(namespace) +
+      ds.mixin.metadata.labels(self.defaultHoneycombLabels) +
+      // Template.
+      ds.mixin.spec.template.metadata.labels(self.defaultHoneycombLabels) +
+      ds.mixin.spec.template.spec.containers(
+        self.agentContainer(containerName, containerTag)) +
+      ds.mixin.spec.template.spec.terminationGracePeriodSeconds(30),
+```
+
+But our arguments aren't defined yet. The next example adds these definitions, plus some 
+other parts that we'll use to assemble higher-level definitions.
+
+Note in particular that the base DaemonSet definition doesn't include RBAC support, which we 
+need, so we add it to the collection of parts:
```

Comment:
I like the structure, I think it will be more effective if we start by showing how to use the parts, rather than their definitions.